feat/davinci integration tests

OndrejTuma opened the pull request:
### Description

Implemented generated workflow for integration tests

### How to test

- CI should be green

### Review

- [ ] Read [CONTRIBUTING.md](https://github.com/toptal/picasso/blob/master/CONTRIBUTING.md) and [Component API principles](https://github.com/toptal/picasso/blob/master/docs/contribution/component-api.md)
- [ ] Annotate all `props` in component with documentation
- [ ] Create `examples` for component
- [ ] Ensure that deployed demo has expected results and good examples
- [ ] Ensure that tests pass by running `yarn test`
- [ ] Ensure that visuals tests pass by running `yarn test:visual`. If not - check the documentation [how to fix visual tests](https://github.com/toptal/picasso/blob/master/docs/contribution/visual-testing.md#fixing-broken-visual-tests-inside-a-pr)
- [ ] Ensure the changed/created components have not caused accessibility issues. [How to use accessibility plugin in storybook](https://github.com/toptal/picasso/blob/master/docs/contribution/accessibility.md).

<details>
<summary>PR commands</summary>
<br />

List of available commands:

- `@toptal-bot run all` - Run whole pipeline
- `@toptal-bot run build` - Check build
- `@toptal-bot run visual` - Run visual tests
- `@toptal-bot run deploy:documentation` - Deploy documentation
- `@toptal-bot run package:alpha-release` - Release alpha version

</details>

<details>
<summary>PR Review Guidelines</summary>
<br />

#### When to approve? ✅

**You are OK** with merging this PR and

1. You have no extra requests.
2. You have optional requests.
   1. Add `nit:` to your comment. (ex. `nit: I'd rename this variable from makeCircle to getCircle`)

#### When to request changes? ❌

**You are not OK** with merging this PR because

1. Something is broken after the changes.
2. Acceptance criteria is not reached.
3. Code is dirty.

#### When to comment (neither ✅ nor ❌)

**You want your comments to be addressed** before merging this PR in cases like:

1. There are leftovers like unnecessary logs, comments, etc.
2. You have an opinionated comment regarding the code that requires a discussion.
3. You have questions.

#### How to handle the comments?

1. An owner of a comment is the only one who can resolve it.
2. An owner of a comment must resolve it when it's addressed.
3. A PR owner must reply with ✅ when a comment is addressed.

</details>

denieler commented on .github/workflows/davinci-integration-tests.yml:
```diff
@@ -0,0 +1,170 @@
+name: Davinci Integration tests
```
why do we need to have `Davinci` keyword here and in the file name? I would just remove it 🤷‍♂️

OndrejTuma replied:
Sure I can do that 👍 probably good idea to remove it also from skeleton file

denieler commented on .github/workflows/davinci-integration-tests.yml:
```diff
@@ -0,0 +1,170 @@
+        uses: actions/checkout@v2
+
+      - name: Setup node
+        uses: actions/setup-node@v2
```
which version of node it uses?

OndrejTuma replied:
The one that is installed on the runner by default, in ubuntu-latest version its Node 16.14.2

denieler commented on .github/workflows/davinci-integration-tests.yml:
```diff
@@ -0,0 +1,170 @@
+        if: ${{ env.PRINT_COVERAGE == 'true' }}
+        run: |
+          mkdir integration-coverage
+          mv $GROUP_LOCATION/$COVERAGE_REPORT_DIR/coverage-final.json integration-coverage/coverage.${GROUP_PACKAGE////-}.$GROUP_INDEX.json
```
what does this `////-`?

OndrejTuma replied:
Just replaces `/` in the package name with `-`

denieler replied:
looks a bit weird 😃

OndrejTuma replied:
it does 🙂  But I think in our use-case GROUP_PACKAGE is `default` so this step can be simplified

OndrejTuma replied:
I removed it altogether

denieler commented on .github/workflows/davinci-integration-tests.yml:
```diff
@@ -0,0 +1,170 @@
+        env:
+          PARALLEL_GROUPS: ${{ matrix.total }}
+        with:
+          command: yarn ${{ env.COMMAND }}
```
I would write here `test:integration:ci`

OndrejTuma replied:
Why not 😁  I already changed the auto-generated file quite a lot

denieler commented on .github/workflows/davinci-integration-tests.yml:
```diff
@@ -0,0 +1,170 @@
+name: Davinci Integration tests
+
+on:
+  pull_request:
+    types: [opened, synchronize, reopened, edited]
+    branches: [master, main]
```
I would like to have all pr checks in the same workflow file - `pr.yaml`

OndrejTuma replied:
But what about integration tests in `visual-testing.yml` workflow?

OndrejTuma replied:
I mean it would be faster I think, but since integration test WF has a lot of jobs, it would be hard to read wouldnt it?

OndrejTuma replied:
I created a callable workflow so all checks are technically part of one workflow, but the workflow is separated, so that it can be called from release workflow as well

denieler reviewed (changes requested):
I think we should have one single workflow file for pr checks - `pr.yaml`

denieler commented on .github/workflows/davinci-integration-tests.yml:
```diff
@@ -0,0 +1,170 @@
+        if: ${{ env.PRINT_COVERAGE == 'true' }}
+        uses: ./.github/actions/report-coverage
+        with:
+          name: integration-coverage
```
why not move those steps to `pr.yaml` file?

OndrejTuma replied:
I'll see what I can do 🖖

OndrejTuma commented:
> I think we should have one single workflow file for pr checks - `pr.yaml`

What about `visual-testing` workflow that does the same thing when PR is merged to master @denieler ?

OndrejTuma commented:
> > I think we should have one single workflow file for pr checks - `pr.yaml`
> 
> What about `visual-testing` workflow that does the same thing when PR is merged to master @denieler ?

@denieler So I reworked it to a called workflow and simplified it. Tell me what you think 🙂

denieler commented on cypress/plugins/index.js:
```diff
@@ -17,6 +23,8 @@ module.exports = (on, config) => {
   // add React preset to be able to transpile JSX
   babelLoader.options.presets.push(require.resolve('@babel/preset-react'))
+  // add istanbul plugin for instrumenting code
+  babelLoader.options.plugins = [require.resolve('babel-plugin-istanbul')]
```
```suggestion
  babelLoader.options.plugins = [
      ...babelLoader.options.plugins,
      require.resolve('babel-plugin-istanbul')
  ]
```

OndrejTuma replied:
plugins are undefined for `babelLoader.options`, so there's no need

denieler replied:
I know, but in the future we can add some plugins there, so we would not need to fix then

OndrejTuma replied:
OK makes sense 👍  thank you for explaining 🙇

OndrejTuma replied:
Oh but I actually cannot iterate over `babelLoader.options.plugins` because its undefined 😬  so I changed it a little bit

denieler commented on .github/workflows/pr.yaml:
```diff
@@ -18,7 +18,8 @@ concurrency:
 
 env:
-  TOPTAL_DEVBOT_TOKEN: ${{ secrets.TOPTAL_DEVBOT_TOKEN }}
+  GITHUB_TOKEN: ${{ secrets.TOPTAL_DEVBOT_TOKEN }}
```
I would not do that, because it's becoming confusing whether it's a default internal `GITHUB_TOKEN` token or it's a custom one. They behave completely different in GH Actions. So I would keep the name

OndrejTuma replied:
That makes sense 🙂  I've reverted it

denieler commented on .github/workflows/visual-testing.yml:
```diff
@@ -53,11 +53,11 @@ jobs:
-        env:
-          HAPPO_API_KEY: ${{ secrets.HAPPO_API_KEY }}
-          HAPPO_API_SECRET: ${{ secrets.HAPPO_API_SECRET }}
+  integration-tests:
```
it's confusing that `visual-testing` workflow executes integration tests 🤷‍♂️ If you want to have some checks after PR has been merged - I would suggest to rename `visual-testing` workflow to `release.yml` or `master.yml`

denieler commented on .github/workflows/davinci-integration-tests.yml:
```diff
@@ -0,0 +1,106 @@
+
+
+env:
+  GITHUB_TOKEN: ${{ secrets.TOPTAL_DEVBOT_TOKEN }}
```
same here, it's better to use custom name

denieler commented on .github/workflows/davinci-integration-tests.yml:
```diff
@@ -0,0 +1,106 @@
+      - uses: ./.github/actions/yarn-install
+
+      - name: Finalize Happo
+        run: npx happo-e2e finalize
```
in happo docs it says that we need also to use `HAPPO_NONCE` env variable set to all our builds - https://docs.happo.io/docs/cypress#parallel-builds. It looks like we need to add it

OndrejTuma replied:
It is there in the workflow level 👍  I put it there because not only this job but also parallel instances need HAPPO_NONCE and access tokens

denieler replied:
@OndrejTuma I just don't see this variable set in this PR and in the `release.yml` as well. Do I look into wrong direction?

OndrejTuma replied:
@denieler its part of integration-tests workflow: https://github.com/toptal/picasso/pull/2667/files#diff-ed9daa5b9c8d4786687ec281988d936bdca0f6ad192416aa73f9e772c16a24f2R20
I defined it as a simple env variable

denieler replied:
ah, nice, thank you 🙇 @OndrejTuma

TomasSlama reviewed (approved):
Nice that you figure out how to use happo after all parallel tests have finished 👍 ,

TomasSlama commented:
also is this expected? 
![image](https://user-images.githubusercontent.com/6830426/167797550-ae267a7a-d96c-4d86-a3fa-3b89b9813434.png)

augustobmoura commented on cypress/plugins/index.js:
```diff
@@ -2,8 +2,14 @@ const path = require('path')
 const webpackPreprocessor = require('@cypress/webpack-preprocessor')
 const { startDevServer } = require('@cypress/webpack-dev-server')
 const happoTask = require('happo-cypress/task')
+// eslint-disable-next-line import/no-extraneous-dependencies
+const plugins = require('@toptal/davinci-qa/src/configs/cypress/plugins')
```
I think that if we depend on this project directly, we should add it to the `package.json` or re-export it from the main `davinci` package. Wdyt?

So far the `@toptal/davinci-qa` has been considered an internal project only, and it shouldn't be required directly on other projects, so this eslint warning kinda makes sense

OndrejTuma replied:
@augustobmoura davinci has davinci-qa as its dependency, so we can safely import this plugins file in picasso. If we add davinci-qa to package.json explicitly, we would have two versions in yarn lock (or we would have to sync them after each davinci upgrade)
